src/app/ui/editor/pixels_movement.cpp

```diff
@@ -789,32 +849,106 @@ void PixelsMovement::stampImage(bool finalStamp)
   if (currentCel && currentCel->layer() &&
       currentCel->layer()->isImage() &&
       !currentCel->layer()->isEditableHierarchy()) {
-    Transformation initialCelPos(gfx::Rect(m_initialMask0->bounds()), m_currentData.cornerThick());
+    Transformation initialCelPos(gfx::Rect(m_initialMask0->bounds()),
+                                 m_currentData.cornerThick());
     redrawExtraImage(&initialCelPos);
     stampExtraCelImage();
   }
 
+  // Saving original values before the 'for' loop and the
+  // 'reproduceAllTransformationsWithInnerCmds' function for restoring later.
+  // All values of m_initialXX, m_currentXX will be recalculated
+  // to align their original selection bounds with each cel's grid.
+  TilemapMode originalSiteTilemapMode = (
+    m_site.tilemapMode() == TilemapMode::Tiles &&
+    m_site.layer()->isTilemap()? TilemapMode::Tiles : TilemapMode::Pixels);
+  TilesetMode originalSiteTilesetMode = m_site.tilesetMode();
+  auto initialMask0 = std::make_unique<Mask>(*m_initialMask0);
+  auto initialMask  = std::make_unique<Mask>(*m_initialMask);
+  auto currentMask  = std::make_unique<Mask>(*m_currentMask);
+  auto initialData  = std::make_unique<Transformation>(m_initialData);
+  auto currentData  = std::make_unique<Transformation>(m_currentData);
+
```

Comment:
Sometimes `Mask` can be just a value (not a pointer), the same for `Transformation`:
```suggestion
  Mask initialMask(*m_initialMask0);
  Mask initialMask(*m_initialMask);
  Mask currentMask(*m_currentMask);
  auto initialData = m_initialData;
  auto currentData = m_currentData;

```